[BUGFIX] [ENHANCEMENT] Enable Getting a Single Batch from an SQL Database by Specifying an Arbitrary SQL Query -- Part 1: The Essentials

alexsherstinsky opened the pull request:
### Scope
In the Legacy architecture, it was possible to get a batch of data from an SQL datasource by specifying an SQL query as a flexible attribute.  The present enhancement makes it possible to accomplish the same in the current (v0.13) architecture.

#### Usage Examples

The SQL query is supplied via the `batch_data` attribute when specifying the `batch_request` in association with the `RuntimeDataConnector` and `SqlAlchemyExecutionEngine`.

Formally:
```
    batch = context.get_batch(
        batch_request=BatchRequest(
            datasource_name="my_runtime_sql_datasource",
            data_connector_name="my_runtime_data_connector",
            data_asset_name="IN_MEMORY_DATA_ASSET",
            batch_data="SELECT * FROM table_partitioned_by_date_column__A",
            partition_request=PartitionRequest(
                **{
                    "partition_identifiers": {
                        "pipeline_stage_name": "core_processing",
                        "airflow_run_id": 1234567890,
                    },
                },
            ),
        ),
    )
```

or (without the typed `BatchRequest` object):
```
    batch = context.get_batch(
        datasource_name="my_runtime_sql_datasource",
        data_connector_name="my_runtime_data_connector",
        data_asset_name="IN_MEMORY_DATA_ASSET",
        batch_data="SELECT * FROM table_partitioned_by_date_column__A",
        partition_request={
            "partition_identifiers": {
                "pipeline_stage_name": "core_processing",
                "airflow_run_id": 1234567890,
            },
        },
    )
```

### Proposed Next Immediate Clean-Up Items
* Hat tip @eugmandel: We might wish to turn `batch_data` from type `Any` to type `dict` so as to allow more intuitive way to specify the source of the data.  For example:
```
batch_data = {
    "dataframe": my_spark_df,
}
```

or
```
batch_data = {
    "query": "SELECT title, genre FROM movies LIMIT 100",
}
```

(with appropriate error checking).
* We need to finally reconcile `asset` and `data_asset` into a single, backward-compatible entity: `data_asset`.  Maintaining both is making the code more difficult to understand (and only one, `data_asset`, should be needed for the design).
* In `batch.py`, we define `to_json_dict()` and `get_json_dict()` -- only one, `to_json_dict()`, which implements the interface `SerializableDictDot.to_json_dict()`, should be needed, since their logic is the same.
* A few more tests insuring that `batch_spec` supplied to `ExecutionEngine` has the correct type would be helpful.


Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-


After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


Thank you for submitting!

Commit:
LINT

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Essentials of supporting an SQL query via RuntimeDataConnector and SqlAlchemyExecutionEngine.

Commit:
Test compatibility management.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Shinnnyshinshin commented on great_expectations/execution_engine/sqlalchemy_execution_engine.py:
```diff
@@ -1035,20 +1040,37 @@ def _build_selectable_from_batch_spec(self, batch_spec):
+        )
+        if isinstance(batch_spec, RuntimeDataBatchSpec):
+            # batch_data != None is already checked when RuntimeDataBatchSpec is instantiated
+            query: str = batch_spec.batch_data
```
the comment is very helpful :) thanks

Shinnnyshinshin commented on tests/conftest.py:
```diff
@@ -2333,6 +2333,30 @@ def titanic_pandas_data_context_with_v013_datasource_with_checkpoints_v1_with_em
         """
 
     context.test_yaml_config(name="my_datasource", yaml_config=datasource_config)
+
```
I'm seeing a number of tests that use this feature are currently failing... would it help to split it out into two sets of fixtures? 
ie. one that lives in `tests/data_context/conf.py` and another that lives in `tests/checkpoint/conf.py`?

alexsherstinsky replied:
Just fixed this!  We now have an SQL Execution Engine in one of the fixtures -- this required a more careful checking.

Shinnnyshinshin replied:
Ok wonderful to hear @alexsherstinsky  :) Will defer to @jcampbell for final approval

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Make tests work properly when SQL Alchemy is not installed.

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Update compatibility.

Commit:
Compatibility.

Commit:
Compatibility.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Resolve merge conflicts.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Clean up.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Lint.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
merge

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Minor cleanup.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim commented on tests/execution_engine/test_sqlalchemy_batch_data.py:
```diff
@@ -35,7 +35,26 @@ def test_instantiation_with_table_name(sqlite_view_engine):
+
+def test_instantiation_with_selectable():
+    # Note Abe 20111119: Fill this in
+    pass
```
Where did this come from? Weird that I don't see it in `develop`.

alexsherstinsky replied:
It is not there any more, and I will delete it.  Must have been in `develop` before, and someone else deleted it.

roblim commented on tests/execution_engine/test_sqlalchemy_batch_data.py:
```diff
@@ -102,7 +124,7 @@ def test_instantiation_with_and_without_temp_table(sqlite_view_engine, sa):
     selectable = sa.select("*").select_from(sa.table(name="test_table", schema="main"))
     SqlAlchemyBatchData(
-        execution_engine=engine,
+        execution_engine=execution_engine,
```
I believe `engine` on line 142 needs to be changed to `execution_engine`. Couldn't add comment on actual line, since it wasn't part of PR, so adding comment here.

alexsherstinsky replied:
There is no line 142 there any more...  And it says `execution_engine` everywhere.  I know that `GitHub` is not always clear about the changes, and sometimes you one has to go to the working branch to be sure.  Happy to pair on this.  Thanks.

alexsherstinsky replied:
Ooo!  I found one `engine` and fixed it.  Thanks.  (It was on line 139.)

roblim reviewed (changes requested):
Looks great! Just a few minor things - see comments. Per last week's design review meeting, I think the only thing that needs to be added is moving `partition_identifiers` up one level, to be a top-level key/arg.

spbail commented:
> Looks great! Just a few minor things - see comments. Per last week's design review meeting, I think the only thing that needs to be added is moving `partition_identifiers` up one level, to be a top-level key/arg.

We decided to not move these for now, but only implement the `batch_data` change. We'll do another pass over the design tomorrow to discuss the partition_identifiers change in more detail. @jcampbell correct me if I misunderstood that?

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

alexsherstinsky commented:
> > Looks great! Just a few minor things - see comments. Per last week's design review meeting, I think the only thing that needs to be added is moving `partition_identifiers` up one level, to be a top-level key/arg.
> 
> We decided to not move these for now, but only implement the `batch_data` change. We'll do another pass over the design tomorrow to discuss the partition_identifiers change in more detail. @jcampbell correct me if I misunderstood that?

@roblim I agree with @spbail -- at least from the perspective that moving `partition_identifiers` will not be a quick change on the account of having to carefully update all the `BatchRequest` parsing code, usage in the various `DataConnector` classes, and in a large number of tests across multiple areas (`DataConnector`, `DataSource`, and `DataContext`).  Thanks.

Commit:
Follow up on comments.

Commit:
merge

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

Commit:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Rename partition_identifiers to batch_identifiers

roblim committed:
Change “runtime_keys” to “batch_identifiers”

roblim committed:
Linting

roblim committed:
Nix batch_data top-level attribute on BatchRequest and BatchRequestBase

roblim committed:
Add RuntimeBatchRequest

roblim committed:
Add query and runtime_parameters to get_batch_list; use RuntimeBatchRequest

roblim committed:
Add RuntimeQueryBatchSpec

roblim committed:
Add runtime_parameters and batch_identifiers attributes to BatchRequestBase

roblim committed:
Update to_json method

roblim committed:
Nix get_json_dict in favor of updating method on BatchRequestBase

roblim committed:
Add runtime_paramters validation

roblim committed:
Move properties to BatchRequestBase

roblim committed:
Nix limit arg

roblim committed:
Fix typo

roblim committed:
Add FIXME

roblim committed:
Nix top-level batch_data and move into runtime_parameters

roblim committed:
Return appropriate batch_spec according to runtime_parameters

roblim committed:
Nix top-level PartitionReqquest in favor of top-level batch_identifiers

roblim committed:
Use batch_request data_asset_name if present

roblim committed:
Change expected type to RuntimeBatchRequest

roblim committed:
_validate_batch_request is no longer needed

- runtime_parameters is the particular attribute on the RuntimeBatchRequest that would need validation, but this already happens on instantiation of RuntimeBatchRequest

roblim committed:
Validate top-level batch_identifiers arg if present

roblim committed:
Also allow RuntimeBatchRequest type

roblim committed:
Use runtime_parameters instead

roblim committed:
Check for RuntimeBatchRequest type instead of batch_data key

roblim committed:
Add InvalidBatchRequestError

roblim committed:
Use new RuntimeQueryBatchSpec in conditional check

roblim committed:
Update fixture name

roblim committed:
Update fixture name

roblim committed:
Make it more clear that data_asset_name is user-generated

roblim committed:
Update test to reflect new runtime_parameters arg

roblim committed:
Nix dupliate code and update test

roblim committed:
Linting

roblim committed:
Also allow RuntimeBatchRequest

roblim committed:
Add path

roblim committed:
Nix limit

roblim committed:
Isort and linting

roblim committed:
Update tests

roblim committed:
Parse path to determine type of BatchSpec to return

roblim committed:
Isort and linting

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

# Conflicts:
#	tests/profile/test_user_configurable_profiler_v3_batch_request.py

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Use `is not None`

roblim committed:
Add support for RuntimeBatchRequest

roblim committed:
Add RuntimeBatchRequest

roblim committed:
Update tests

roblim committed:
Update tests

roblim committed:
Update test

roblim committed:
Update conditional to include RuntimeQueryBatchSpec

roblim committed:
Move runtime_parameter validation to RuntimeDataConnector

- restore _validate_batch_request method on RuntimeDataConnector and update

roblim committed:
Also make sure runtime_parameters and batch_identifiers are non-empty

roblim committed:
Revert change allowing custom data_asset_name

roblim committed:
Update tests

roblim committed:
Update test

roblim committed:
Update test

roblim committed:
Update test

roblim committed:
Move/update tests

roblim committed:
Fix typo

roblim committed:
Add tests for path in runtime_parameters

roblim committed:
Switch to any

roblim committed:
Isort and linting

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Nix unused imports

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

roblim committed:
Merge branch 'develop' into alexsherstinsky/bugfix/enable_passing_sql_query_via_runtime_data_connector_essentials-02_17_2021_41

# Conflicts:
#	great_expectations/datasource/data_connector/runtime_data_connector.py
#	tests/datasource/data_connector/test_runtime_data_connector.py
#	tests/datasource/test_new_datasource.py

roblim committed:
Update tests

roblim committed:
Update tests

roblim committed:
Update tests

roblim committed:
Isort and linting

roblim commented:
Closing in favor of #2550
